plugins/ua_network_pubsub_ethernet.c

```diff
@@ -0,0 +1,422 @@
+    channelDataEth->prio = (UA_Byte) ((ptrPrio == NULL) ? 0 : atoi(ptrPrio));
+
+    return UA_STATUSCODE_GOOD;
+}
+
+/*****************************************************************************
+ * Plugin Functions
+ ****************************************************************************/
+
+/**
+ * Open communication socket based on the connectionConfig.
+ *
+ * @return ref to created channel, NULL on error
+ */
+static UA_PubSubChannel *
+UA_PubSubChannelEthernet_open(const UA_PubSubConnectionConfig *connectionConfig) {
+
+    UA_LOG_INFO(UA_Log_Stdout, UA_LOGCATEGORY_SERVER,
+                "Open PubSub ethernet connection.");
+
+    /* allocate and init memory for the ethernet specific internal data */
+    UA_PubSubChannelDataEth* channelDataEth =
+            (UA_PubSubChannelDataEth*) UA_malloc(sizeof(*channelDataEth));
+    if (!channelDataEth) {
+        UA_LOG_ERROR (UA_Log_Stdout, UA_LOGCATEGORY_SERVER,
+            "PubSub Connection creation failed. Out of memory.");
+        return NULL;
+    }
+    memset(channelDataEth, 0, sizeof(*channelDataEth));
+
+    /* handle specified network address */
+    UA_NetworkAddressUrlDataType address;
+    if (UA_Variant_hasScalarType (&connectionConfig->address,
+                                 &UA_TYPES[UA_TYPES_NETWORKADDRESSURLDATATYPE])) {
+        address = *(UA_NetworkAddressUrlDataType *) connectionConfig->address.data;
+    } else {
+        UA_LOG_ERROR (UA_Log_Stdout, UA_LOGCATEGORY_SERVER,
+            "PubSub Connection creation failed. Invalid Address.");
+        UA_free(channelDataEth);
+        return NULL;
+    }
+    UA_LOG_DEBUG(UA_Log_Stdout, UA_LOGCATEGORY_SERVER, "Specified Interface Name = %.*s",
+         (int) address.networkInterface.length, address.networkInterface.data);
+    UA_LOG_DEBUG(UA_Log_Stdout, UA_LOGCATEGORY_SERVER, "Specified Network Url = %.*s",
+         (int)address.url.length, address.url.data);
+
+    snprintf(channelDataEth->interfaceName,
+             sizeof(channelDataEth->interfaceName), "%.*s",
+             (int) address.networkInterface.length,
+             address.networkInterface.data);
+
+    /* encode the URL and store information in internal structure */
+    if (UA_parseEndpointUrlEthernet(&address.url, channelDataEth)) {
+        UA_LOG_ERROR (UA_Log_Stdout, UA_LOGCATEGORY_SERVER,
+            "PubSub Connection creation failed. Invalid Address URL.");
+        UA_free(channelDataEth);
+        return NULL;
+    }
+
+    /* generate a new Pub/Sub channel and open a related socket */
+    UA_PubSubChannel *newChannel;
+    newChannel = (UA_PubSubChannel*)UA_calloc(1, sizeof(UA_PubSubChannel));
+    if (!newChannel) {
+        UA_LOG_ERROR(UA_Log_Stdout, UA_LOGCATEGORY_SERVER, "PubSub Connection creation failed. Out of memory.");
+        UA_free(channelDataEth);
+        return NULL;
+    }
+
+    /* open a RAW socket */
+    int sockFd;
+    if ((sockFd = UA_socket(PF_PACKET, SOCK_RAW, htons(ETH_P_ALL))) < 0) {
```

Comment:
by using ETH_P_ALL all packets will be received regardless of their ethertype.. i guess this is not the intended behaviour.

I see that there are two ethertypes needed ETHERTYPE_UADP and ETHERTYPE_VLAN. Anybody has any idea how to solve this? Is it possible to bind one raw socket to two ethertypes? use two sockets? use ETH_P_ALL with a custom BPF? All of this is really linux dependant.

Btw i don't know if receiving raw packets with etherype=VLAN is working with network cards which supports hardware vlan filtering, so this should be tested.

Reply:
OPC UA PubSub uses a custom ethertype.